### PR TITLE
add is quadratic test

### PR DESCRIPTION
**Pull request by matklad**
bors r+
🤖

### CODE REVIEW

File: crates/ide/src/syntax_highlighting/tests.rs [hunk 1]
```diff
@@ -257,6 +260,86 @@ fn benchmark_syntax_highlighting_long_struct() {
...
+        for (x, y) in xy.clone() {
+            num += (x - mean_x) * (y - mean_y);
+            denom += (x - mean_x).powi(2);
+        }
+        num / denom
+    };
+
+    let a = mean_y - b * mean_x;
+
+    let mut plot = format!("y_pred = {:.3} + {:.3} * x\n\nx y y_pred\n", a, b);
```
Comment (DJMcNab) [inline crates/ide/src/syntax_highlighting/tests.rs:316]: I guess a table is a plot for some definition thereof 🤔

**Comment by matklad** [comment]
bors try

**Comment by matklad** [comment]
changelog internal: use liner regression to catch "accidentally quadratic" performance regressions.

**Comment by matklad** [comment]
bors try+

**Comment by matklad** [comment]
bors try

**Comment by matklad** [comment]
bors try+

**Comment by matklad** [comment]
bors try

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -76,7 +76,7 @@ jobs:
       run: cargo test --no-run --locked
 
     - name: Test
-      run: cargo test
+      run: cargo test -- --nocapture
```
Comment (lnicola) [inline .github/workflows/ci.yaml:79]: Do we still need this?

File: crates/ide/src/syntax_highlighting/tests.rs [hunk 3]
```diff
@@ -257,6 +260,99 @@ fn benchmark_syntax_highlighting_long_struct() {
...
+
+    // Linear regression: finding a and b to fit y = a + b*x.
+
+    let mean_x = mean(&xs);
+    let mean_y = mean(&ys);
+
+    let b = {
+        let mut num = 0.0;
+        let mut denom = 0.0;
+        for (x, y) in xy.clone() {
```
Comment (lnicola) [inline crates/ide/src/syntax_highlighting/tests.rs:320]: That's a lot of `zip`ping and `unzip`ping :smile:.

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (matklad) [reply .github/workflows/ci.yaml:79]: Yup! I believe tests shouldn't print anything by default anyway. I'd say it's even a cargo bug that `--nocapture` is a thing at all :-)

**Comment by matklad** [comment]
cc @SomeoneToIgnore this is one of the reasons want to try using generated code as much as possible. With generation, it's easy to vary `n` and to check big-O

File: crates/ide/src/syntax_highlighting/tests.rs [hunk 4]
```diff
@@ -257,6 +260,99 @@ fn benchmark_syntax_highlighting_long_struct() {
...
+        se += (y - y_pred).powi(2);
+        max_error = max_error.max((y_pred - y).abs());
+
+        format_to!(plot, "{:.3} {:.3} {:.3}\n", x, y, y_pred);
+    }
+
+    let rmse = (se / xs.len() as f64).sqrt();
+    format_to!(plot, "\nrmse = {:.3} max error = {:.3}", rmse, max_error);
+
+    assert!(rmse < 0.05 && max_error < 0.1 && a > -0.1, "\nLooks quadratic\n{}", plot);
```
Comment (DJMcNab) [inline crates/ide/src/syntax_highlighting/tests.rs:344]: 'looks nonlinear' might be a better message here

**Pull request merged**